src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -40,22 +40,176 @@
+ *                                                   checks for both. When further unrolling this loop, we only need to
+ *                                                   keep the checks on the first and last access (e.g. a[i] and a[i+3]
+ *                                                   on the next unrolling step as they cover the checks in the middle
+ *                                                   for a[i+1] and a[i+2]).
+ *                                                   Therefore, we just need to cover:
+ *                                                   - Initial value: a[init]
+ *                                                   - Last value: a[init + new stride - original stride]
+ *                                                   (We could still only use one Template Assertion Predicate to create
+ *                                                   both Initialized Assertion Predicates from - might be worth doing
+ *                                                   at some point).
+ *                                                   When later splitting a loop (pre/main/post, peeling, unrolling),
+ *                                                   we create two Initialized Assertion Predicates from the Template
+ *                                                   Assertion Predicates by replacing the OpaqueLoop* nodes by actual
+ *                                                   values. Initially (before unrolling), both Assertion Predicates are
+ *                                                   equal. The Initialized Assertion Predicates are always true because
+ *                                                   their range is covered by a corresponding Hoisted Predicate.
+ *                        - Range Check Elimination: A range check is removed from the main-loop by changing the pre
+ *                                                   and main-loop iterations. We add two additional Template Assertion
+ *                                                   Predicates (see explanation in section above) and one Initialized
+ *                                                   Assertion Predicate for the just removed range check. When later
+ *                                                   unrolling the main-loop, we create two Initialized Assertion
+ *                                                   Predicates from the Template Assertion Predicates by replacing the
+ *                                                   OpaqueLoop* nodes by actual values for the unrolled loop.
+ *                                                   The Initialized Assertion Predicates are always true because we will
+ *                                                   never enter the main loop because of the changed pre- and main-loop
+ *                                                   exit conditions.
```

Comment:
This does still not quite sound right. We will never enter the main loop? Sounds like the main-loop is ueseless in all cases. Suggestion:

```
The Initialized Assertion Predicates are always true: they are true when we enter the main loop
(because we adjusted the pre-loop exit condition), they are true in the last iteration (because we
adjust the main-loop exit condition), and they are true in all iterations in the middle by implication.
```

Reply:
Thanks for the suggestion, I've pushed an update.